src/main.rs

```diff
@@ -1,13 +1,186 @@
+            .arg(
+                Arg::with_name("protocol")
+                    .short("p")
+                    .long("protocol")
+                    .help("select transport protocol")
+                    .takes_value(true)
+                    .default_value("http"),
+            )
+            .arg(
+                Arg::with_name("endpoint")
+                    .short("e")
+                    .long("endpoint")
+                    .help("http port")
+                    .takes_value(true)
+                    .default_value("7878"),
+            )
+            .arg(
+                Arg::with_name("pool_size")
+                    .short("s")
+                    .long("pool-size")
+                    .help("max size of threadpool")
+                    .takes_value(true)
+                    .default_value("2"),
+            )
+            .get_matches_from_safe(args)
+            .unwrap_or_else(|e| panic!("Unable to parse args: {}", e));
+
+        println!("Launching Trin...");
+        let protocol = matches.value_of("protocol").unwrap();
+        let endpoint = matches.value_of("endpoint").unwrap();
+        let endpoint = match endpoint.parse::<u32>() {
+            Ok(n) => n,
+            Err(_) => panic!("Provided endpoint arg is not a number"),
+        };
+        let pool_size = matches.value_of("pool_size").unwrap();
+        let pool_size = match pool_size.parse::<u32>() {
+            Ok(n) => n,
+            Err(_) => panic!("Provided pool size arg is not a number"),
+        };
+
+        // parse protocol & endpoint
+        match protocol {
+            "http" => println!("Protocol: {}\nEndpoint: {}", protocol, endpoint),
+            "ipc" => match endpoint {
+                7878 => println!("Protocol: {}", protocol),
+                _ => panic!("No ports for ipc connection"),
+            },
+            val => panic!(
+                "Unsupported protocol: {}, supported protocols include http & ipc.",
+                val
+            ),
+        }
+        println!("Pool Size: {}", pool_size);
+
+        let infura_project_id = match env::var("TRIN_INFURA_PROJECT_ID") {
+            Ok(val) => val,
+            Err(_) => panic!(
+                "Must supply Infura key as environment variable, like:\n\
+                TRIN_INFURA_PROJECT_ID=\"your-key-here\" trin"
+            ),
+        };
+
+        Ok(TrinConfig {
+            endpoint: endpoint,
+            infura_project_id: infura_project_id,
+            pool_size: pool_size,
+            protocol: protocol.to_string(),
+        })
+    }
+}
+
 fn main() {
-    match env::var("TRIN_INFURA_PROJECT_ID") {
-        Ok(val) => cli::launch_trin(val),
-        Err(_) => println!(
-            "Must supply Infura key as environment variable, like:\n\
-            TRIN_INFURA_PROJECT_ID=\"your-key-here\" trin"
-        ),
+    let trin_config = TrinConfig::new();
+    cli::launch_trin(trin_config);
+}
+
+#[cfg(test)]
```

Comment:
From what I've read - to do proper unit testing we'll need to pull out the majority of the logic from the binary (`main.rs`) into the library (`lib.rs`) and just replace `main.rs` with a thin wrapper around the library logic. I've just included the tests here for now, but can do the re-organizing in this or a separate pr.

Reply:
Yeah, I think all of the `impl TrinConfig` can go into the `cli.rs` and be tested there, leaving only the `main()` here.